smolyay/basis.py

```diff
@@ -403,29 +482,393 @@ def _derivative(self, x):
         float
             Value of the derivative of Trigonometric polynomial.
 
+        n : int, optional
+            order of derivative. Default is 1.
+
         Raises
         ------
         ValueError
             If input is outside the domain `[0, 2\pi]`
         """
         x = numpy.asarray(x)
-        return self.frequency * 1j * numpy.exp(x * self.frequency * 1j)
+        return numpy.exp(x * self.frequency * 1j) * (self.frequency * 1j) ** n
+
+
+class BasisFunctionSet(collections.abc.Sequence):
+    """Set of basis functions and sample points."""
+
+    @abc.abstractmethod
+    def __init__(self):
+        pass
+
+    @property
+    def basis_functions(self):
+        """list: Basis functions."""
+        return self._basis_functions
+
+    @property
+    def domain(self):
+        """numpy.ndarray: Domain of the `basis_functions`"""
+        return self.basis_functions[0].domain
 
+    def __len__(self):
+        return len(self.basis_functions)
 
-class BasisFunctionSet:
-    """Set of basis functions and sample points.
+    def __getitem__(self, key):
+        return self.basis_functions[key]
+
+    def _scale_to_domain(self, points, old_domain):
+        points = self.domain[0] + (self.domain[1] - self.domain[0]) * (
+            (points - old_domain[0]) / (old_domain[1] - old_domain[0])
+        )
+        return numpy.clip(points, self.domain[0], self.domain[1])
+
+    def __call__(self, X, X_domain):
+        """Evaluate all the basis functions in the set
+
+        Calls all the basis function(s) at index and evaluates at X.
+
+        Parameters
+        ----------
+        X : array-like
+            the points to evaluate.
+
+        X_domain : numpy array of shape (2,)
+            the lower and upper bounds of X.
+
+        Returns
+        -------
+        scalar or ndarray
+            the values of the basis functions.
+        """
+
+        new_X = self._scale_to_domain(numpy.asarray(X), X_domain)
+        if any(bf._is_complex for bf in self):
+            y = numpy.zeros([len(self)] + list(new_X.shape), dtype="complex_")
+        else:
+            y = numpy.zeros([len(self)] + list(new_X.shape))
```

Comment:
I think I made a suggestion related to this in the last review, but to reduce duplication, only put the thing you care about (dtype) in the if. Here it is as a one-liner:
```suggestion
        y = numpy.zeros(
            [len(self)] + list(new_X.shape),
            dtype=complex if any(bf._is_complex for bf in self) else float
        )
```